File: src/lib/protocol/nip19.ts
import { bech32 } from '@scure/base';
import { bytesToHex, hexToBytes } from './bytes';
import {
  decodeTlv,
  encodeTlv,
  hexTlv,
  textTlv,
  tlvHex,
  tlvKind,
  tlvRelays,
  tlvText,
  uint32Bytes,
} from './nip19-tlv';
import type { AddressPointer, EventPointer, NostrEntity } from './nip19-types';

const bech32Limit = 5000;

export type {
  AddressPointer,
  EventPointer,
  NostrEntity,
  ProfilePointer,
} from './nip19-types';

export function decodeEntity(value: string): NostrEntity | undefined {
  if (value.length > bech32Limit) return undefined;
  try {
    const { prefix, words } = bech32.decode(value, bech32Limit);
    const bytes = bech32.fromWords(words);
    if (prefix === 'npub' && bytes.length === 32)
      return { type: 'npub', data: bytesToHex(bytes) };
    if (prefix === 'nsec' && bytes.length === 32)
      return { type: 'nsec', data: bytes };
    if (prefix === 'note' && bytes.length === 32)
      return { type: 'note', data: bytesToHex(bytes) };
    if (prefix === 'nprofile') return decodeNprofile(bytes);
    if (prefix === 'nevent') return decodeNevent(bytes);
    if (prefix === 'naddr') return decodeNaddr(bytes);
    return undefined;
  } catch {
    return undefined;
  }
}

export function encodeNpub(pubkey: string): string {
  return encodeBytes('npub', hexToBytes(pubkey));
}

export function encodeNote(id: string): string {
  return encodeBytes('note', hexToBytes(id));
}

export function encodeNsec(secretKey: Uint8Array | string): string {
  return encodeBytes(
    'nsec',
    typeof secretKey === 'string' ? hexToBytes(secretKey) : secretKey,
  );
}

export function encodeNprofile(input: {
  readonly pubkey: string;
  readonly relays?: readonly string[];
}): string {
  return encodeBytes('nprofile', pointerTlv(0, input.pubkey, input.relays));
}

export function encodeNevent(input: EventPointer): string {
  const records = recordsWithRelays(input.relays);
  records[0] = [hexTlv(input.id)];
  if (input.author) records[2] = [hexTlv(input.author)];
  if (input.kind !== undefined) records[3] = [uint32Bytes(input.kind)];
  return encodeBytes('nevent', encodeTlv(records));
}

export function encodeNaddr(input: AddressPointer): string {
  const records = recordsWithRelays(input.relays);
  records[0] = [textTlv(input.identifier)];
  records[2] = [hexTlv(input.pubkey)];
  records[3] = [uint32Bytes(input.kind)];
  return encodeBytes('naddr', encodeTlv(records));
}

function decodeNprofile(bytes: Uint8Array): NostrEntity | undefined {
  const records = decodeTlv(bytes);
  if (!records) return undefined;
  const pubkey = tlvHex(records[0]?.[0]);
  return pubkey
    ? { type: 'nprofile', data: { pubkey, relays: tlvRelays(records) } }
    : undefined;
}

function decodeNevent(bytes: Uint8Array): NostrEntity | undefined {
  const records = decodeTlv(bytes);
  const id = records ? tlvHex(records[0]?.[0]) : undefined;
  if (!records || !id) return undefined;
  const author = records[2]?.[0] ? tlvHex(records[2][0]) : undefined;
  if (records[2]?.[0] && !author) return undefined;
  const kind = records[3]?.[0] ? tlvKind(records[3][0]) : undefined;
  if (records[3]?.[0] && kind === undefined) return undefined;
  return {
    type: 'nevent',
    data: { id, relays: tlvRelays(records), author, kind },
  };
}

function decodeNaddr(bytes: Uint8Array): NostrEntity | undefined {
  const records = decodeTlv(bytes);
  const identifier = records ? tlvText(records[0]?.[0]) : undefined;
  const pubkey = records ? tlvHex(records[2]?.[0]) : undefined;
  const kind = records ? tlvKind(records[3]?.[0]) : undefined;
  if (!records || identifier === undefined || !pubkey || kind === undefined)
    return undefined;
  return {
    type: 'naddr',
    data: { identifier, pubkey, kind, relays: tlvRelays(records) },
  };
}

function pointerTlv(
  type: number,
  hex: string,
  relays: readonly string[] = [],
): Uint8Array {
  const records = recordsWithRelays(relays);
  records[type] = [hexTlv(hex)];
  return encodeTlv(records);
}

function recordsWithRelays(relays: readonly string[] = []) {
  const records: Record<number, Uint8Array[]> = {};
  if (relays.length > 0) records[1] = relays.map(textTlv);
  return records;
}

function encodeBytes(prefix: string, bytes: Uint8Array): string {
  return bech32.encode(prefix, bech32.toWords(bytes), bech32Limit);
}

File: crates/lkjstr-protocol/src/nip19.rs
use std::fmt;

use serde::{Deserialize, Serialize};

use crate::{bytes_to_hex, nip19_tlv};

const BECH32_LIMIT: usize = 5000;

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct ProfilePointer {
    pub pubkey: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relays: Option<Vec<String>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct EventPointer {
    pub id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relays: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub author: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kind: Option<u64>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct AddressPointer {
    pub identifier: String,
    pub pubkey: String,
    pub kind: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relays: Option<Vec<String>>,
}

#[derive(Clone, Eq, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", content = "data")]
pub enum NostrEntity {
    #[serde(rename = "npub")]
    Npub(String),
    #[serde(rename = "nsec")]
    Nsec(String),
    #[serde(rename = "note")]
    Note(String),
    #[serde(rename = "nprofile")]
    Nprofile(ProfilePointer),
    #[serde(rename = "nevent")]
    Nevent(EventPointer),
    #[serde(rename = "naddr")]
    Naddr(AddressPointer),
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Nip19Error {
    InvalidEntity,
    InvalidPrefix,
    InvalidHex,
    InvalidTlv,
}

impl fmt::Debug for NostrEntity {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Npub(value) => formatter.debug_tuple("Npub").field(value).finish(),
            Self::Nsec(_) => formatter.debug_tuple("Nsec").field(&"<redacted>").finish(),
            Self::Note(value) => formatter.debug_tuple("Note").field(value).finish(),
            Self::Nprofile(value) => formatter.debug_tuple("Nprofile").field(value).finish(),
            Self::Nevent(value) => formatter.debug_tuple("Nevent").field(value).finish(),
            Self::Naddr(value) => formatter.debug_tuple("Naddr").field(value).finish(),
        }
    }
}

pub fn decode_nip19(value: &str) -> Option<NostrEntity> {
    if value.len() > BECH32_LIMIT {
        return None;
    }
    let (hrp, bytes) = bech32::decode(value).ok()?;
    match hrp.as_str().to_ascii_lowercase().as_str() {
        "npub" if bytes.len() == 32 => Some(NostrEntity::Npub(bytes_to_hex(&bytes))),
        "nsec" if bytes.len() == 32 => Some(NostrEntity::Nsec(bytes_to_hex(&bytes))),
        "note" if bytes.len() == 32 => Some(NostrEntity::Note(bytes_to_hex(&bytes))),
        "nprofile" => decode_nprofile(&bytes),
        "nevent" => decode_nevent(&bytes),
        "naddr" => decode_naddr(&bytes),
        _ => None,
    }
}

fn decode_nprofile(bytes: &[u8]) -> Option<NostrEntity> {
    let records = nip19_tlv::decode_tlv(bytes)?;
    let pubkey = nip19_tlv::first_hex(&records, 0)?;
    Some(NostrEntity::Nprofile(ProfilePointer {
        pubkey,
        relays: nip19_tlv::relays(&records),
    }))
}

fn decode_nevent(bytes: &[u8]) -> Option<NostrEntity> {
    let records = nip19_tlv::decode_tlv(bytes)?;
    let id = nip19_tlv::first_hex(&records, 0)?;
    let author = optional_hex(&records, 2)?;
    let kind = optional_kind(&records, 3)?;
    Some(NostrEntity::Nevent(EventPointer {
        id,
        relays: nip19_tlv::relays(&records),
        author,
        kind,
    }))
}

fn decode_naddr(bytes: &[u8]) -> Option<NostrEntity> {
    let records = nip19_tlv::decode_tlv(bytes)?;
    Some(NostrEntity::Naddr(AddressPointer {
        identifier: nip19_tlv::first_text(&records, 0)?,
        pubkey: nip19_tlv::first_hex(&records, 2)?,
        kind: nip19_tlv::first_kind(&records, 3)?,
        relays: nip19_tlv::relays(&records),
    }))
}

fn optional_hex(records: &nip19_tlv::TlvRecords, key: u8) -> Option<Option<String>> {
    if nip19_tlv::has_first(records, key) {
        Some(Some(nip19_tlv::first_hex(records, key)?))
    } else {
        Some(None)
    }
}

fn optional_kind(records: &nip19_tlv::TlvRecords, key: u8) -> Option<Option<u64>> {
    if nip19_tlv::has_first(records, key) {
        Some(Some(nip19_tlv::first_kind(records, key)?))
    } else {
        Some(None)
    }
}
